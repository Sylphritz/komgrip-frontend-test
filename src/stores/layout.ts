import { ref } from 'vue'
import { defineStore } from 'pinia'

export const useLayoutStore = defineStore('layout', () => {
  const sidebarOpened = ref(false)

  const toggleSidebar = () => {
    sidebarOpened.value = !sidebarOpened.value

    if (sidebarOpened.value) {
      document.body.classList.add('g-sidenav-pinned')
    } else {
      document.body.classList.remove('g-sidenav-pinned')
    }
  }

  return { sidebarOpened, toggleSidebar }
})
