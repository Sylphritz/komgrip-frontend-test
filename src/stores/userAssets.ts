import { reactive, ref, watch } from 'vue'
import { defineStore } from 'pinia'
import type { Asset } from '@/api/coincap/types'
import { ASSETS_PER_PAGE } from '@/constants/constants'
import { getAssets, getAssetsByIds } from '@/api/coincap/assets'

export const useUserAssetsStore = defineStore('user-assets', () => {
  // Let's pretend this user has pinned the following assets
  const pinnedAssetIds = reactive(['bitcoin', 'ethereum', 'solana', 'dogecoin'])
  const pinnedAssets = ref<Asset[]>([])
  const fetchingPinnedAssets = ref(false)

  const assets = ref<Asset[]>([])
  const fetching = ref(false)
  const pageOffset = ref(0)

  const getNextPage = () => {
    if (fetching.value) return

    pageOffset.value += ASSETS_PER_PAGE
  }

  const getPreviousPage = () => {
    if (fetching.value) return

    pageOffset.value -= ASSETS_PER_PAGE
  }

  const fetchAssets = async () => {
    if (fetching.value) return

    fetching.value = true
    assets.value = await getAssets(pageOffset.value, ASSETS_PER_PAGE)
    fetching.value = false
  }

  const fetchPinnedAssets = async () => {
    if (fetchingPinnedAssets.value) return

    fetchingPinnedAssets.value = true
    pinnedAssets.value = await getAssetsByIds(pinnedAssetIds)
    fetchingPinnedAssets.value = false
  }

  watch(
    pageOffset,
    async () => {
      await fetchAssets()
    },
    {
      immediate: true
    }
  )

  watch(
    pinnedAssetIds,
    async () => {
      await fetchPinnedAssets()
    },
    {
      immediate: true
    }
  )

  return {
    assets,
    pinnedAssets,
    fetching,
    fetchingPinnedAssets,
    pageOffset,
    getNextPage,
    getPreviousPage
  }
})
