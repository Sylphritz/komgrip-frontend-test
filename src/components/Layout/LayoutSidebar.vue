<template>
  <aside
    class="sidenav navbar navbar-vertical navbar-expand-xs border-0 border-radius-xl my-3 fixed-start ms-3"
    :class="{ 'bg-white': sidebarOpened }"
    id="sidenav-main"
  >
    <SidebarHeader />
    <hr class="horizontal dark mt-0" />
    <div class="collapse navbar-collapse w-auto max-height-vh-100" id="sidenav-collapse-main">
      <ul class="navbar-nav">
        <SidebarItem :icon="ShopIcon" url="/" is-active>Dashboard</SidebarItem>
        <SidebarItem url="#">Menu 1</SidebarItem>
        <SidebarItem url="#">Menu 2</SidebarItem>
        <SidebarItem url="#">Menu 3</SidebarItem>
        <SidebarItem url="#">Menu 4</SidebarItem>
      </ul>
    </div>
    <SidebarFooter />
  </aside>
</template>

<script setup lang="ts">
import SidebarHeader from '@/components/Layout/Sidebar/SidebarHeader.vue'
import ShopIcon from '@/components/Icons/ShopIcon.vue'
import SidebarItem from './Sidebar/SidebarItem.vue'
import SidebarFooter from './Sidebar/SidebarFooter.vue'

import { useLayoutStore } from '@/stores/layout'
import { storeToRefs } from 'pinia'

const layoutStore = useLayoutStore()
const { sidebarOpened } = storeToRefs(layoutStore)
</script>
