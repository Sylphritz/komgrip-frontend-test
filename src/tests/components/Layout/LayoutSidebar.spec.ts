import { describe, it, expect, beforeEach } from 'vitest'

import { mount } from '@vue/test-utils'
import LayoutSidebar from '@/components/Layout/LayoutSidebar.vue'
import { createPinia, setActivePinia } from 'pinia'
import { useLayoutStore } from '@/stores/layout'

describe('LayoutSidebar', () => {
  beforeEach(() => {
    setActivePinia(createPinia())
  })

  it('renders successfully and matches the snapshot', () => {
    const wrapper = mount(LayoutSidebar, {
      shallow: true
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('has a white background when opened in mobile responsive mode', () => {
    const wrapper = mount(LayoutSidebar, {
      shallow: true
    })

    const layoutStore = useLayoutStore()
    layoutStore.sidebarOpened = true

    expect(wrapper.element.classList.contains('bg-white')).toMatchSnapshot()
  })
})
