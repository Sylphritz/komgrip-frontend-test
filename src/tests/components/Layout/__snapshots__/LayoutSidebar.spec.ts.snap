// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LayoutSidebar > has a white background when opened in mobile responsive mode 1`] = `false`;

exports[`LayoutSidebar > renders successfully and matches the snapshot 1`] = `
<aside
  class="sidenav navbar navbar-vertical navbar-expand-xs border-0 border-radius-xl my-3 fixed-start ms-3"
  id="sidenav-main"
>
  <sidebar-header-stub />
  <hr
    class="horizontal dark mt-0"
  />
  <div
    class="collapse navbar-collapse w-auto max-height-vh-100"
    id="sidenav-collapse-main"
  >
    <ul
      class="navbar-nav"
    >
      <sidebar-item-stub
        icon="[object Object]"
        isactive="true"
        url="/"
      />
      <sidebar-item-stub
        isactive="false"
        url="#"
      />
      <sidebar-item-stub
        isactive="false"
        url="#"
      />
      <sidebar-item-stub
        isactive="false"
        url="#"
      />
      <sidebar-item-stub
        isactive="false"
        url="#"
      />
    </ul>
  </div>
  <sidebar-footer-stub />
</aside>
`;
